src/osd/PrimaryLogPG.cc

```diff
@@ -2969,6 +3042,171 @@ void PrimaryLogPG::do_proxy_write(OpRequestRef op, const hobject_t& missing_oid,
+  pwop->ctx = new OpContext(op, m->get_reqid(), pwop->ops, this);
+  pwop->mtime = m->get_mtime();
+
+  ObjectOperation obj_op;
+  bufferlist write_buf;
+  uint64_t ori_offset = m->ops[op_index].op.extent.offset;
+  bufferptr tmp_buf(req_length);
+  write_buf.append(tmp_buf);
+  write_buf.copy_in(0, req_length, m->ops[op_index].indata.c_str() + (req_offset - ori_offset));
+  if (write_buf.length() == 0) {
+    assert(0);
+  }
+  obj_op.add_data(m->ops[op_index].op.op, req_offset % chunk_length, 
+		  req_length, write_buf);
+
+  C_ProxyWrite_Commit *fin = new C_ProxyWrite_Commit(
+      this, ori_soid, get_last_peering_reset(), pwop);
+  ceph_tid_t tid = osd->objecter->mutate(
+    soid.oid, oloc, obj_op, snapc,
+    ceph::real_clock::from_ceph_timespec(pwop->mtime),
+    flags, new C_OnFinisher(fin, &osd->objecter_finisher),
+    &pwop->user_version, pwop->reqid);
+  fin->tid = tid;
+  pwop->objecter_tid = tid;
+  proxywrite_ops[tid] = pwop;
+  in_progress_proxy_ops[ori_soid].push_back(op);
+}
+
+void PrimaryLogPG::do_chunk_read(OpRequestRef op, ObjectContextRef obc, int op_index,
+				 uint64_t req_offset, uint64_t req_length)
+{
+  MOSDOp *m = static_cast<MOSDOp*>(op->get_nonconst_req());
+  object_manifest_t *manifest = &obc->obs.oi.manifest;
+  uint64_t chunk_length = manifest->chunk_map[0].length;
+  uint64_t chunk_index = req_offset - (req_offset % chunk_length);
+
+  if (!manifest->chunk_map.count(chunk_index)) {
+    return;
+  } 
+
+  hobject_t soid = manifest->chunk_map[chunk_index].oid;
+  hobject_t ori_soid = m->get_hobj();
+  object_locator_t oloc(soid);
+  unsigned flags = CEPH_OSD_FLAG_IGNORE_CACHE | CEPH_OSD_FLAG_IGNORE_OVERLAY;
+
+  /* same as do_proxy_read() */
+  flags |= m->get_flags() & (CEPH_OSD_FLAG_RWORDERED |
+			     CEPH_OSD_FLAG_ORDERSNAP |
+			     CEPH_OSD_FLAG_ENFORCE_SNAPC |
+			     CEPH_OSD_FLAG_MAP_SNAP_CLONE);
+
+  dout(10) << __func__ << " Start do chunk proxy read for " << *m 
+	   << " index: " << op_index << " oid: " << soid.oid.name << " req_offset: " << req_offset 
+	   << " req_length: " << req_length << dendl;
+
+  ProxyReadOpRef prdop(std::make_shared<ProxyReadOp>(op, ori_soid, m->ops));
+
+  ObjectOperation *pobj_op = new ObjectOperation;
+  OSDOp &osd_op = pobj_op->add_op(m->ops[op_index].op.op);
+  osd_op.op.extent.offset = req_offset % chunk_length;
+  osd_op.op.extent.length = req_length; 
+
+  ObjectOperation obj_op;
+  obj_op.dup(pobj_op->ops);
+
+  if (pool.info.cache_mode == pg_pool_t::CACHEMODE_WRITEBACK &&
+      (agent_state && agent_state->evict_mode != TierAgentState::EVICT_MODE_FULL)) {
+    for (unsigned i = 0; i < obj_op.ops.size(); i++) {
+      ceph_osd_op op = obj_op.ops[i].op;
+      switch (op.op) {
+	case CEPH_OSD_OP_READ:
+	case CEPH_OSD_OP_SYNC_READ:
+	case CEPH_OSD_OP_SPARSE_READ:
+	case CEPH_OSD_OP_CHECKSUM:
+	  op.flags = (op.flags | CEPH_OSD_OP_FLAG_FADVISE_SEQUENTIAL) &
+		       ~(CEPH_OSD_OP_FLAG_FADVISE_DONTNEED | CEPH_OSD_OP_FLAG_FADVISE_NOCACHE);
+      }
```

Comment:
Once we consider the promote case, we need to figure out how to trigger the ref-counting operations that happen when we clear a manifest.  This will require a bit of thinking.  Maybe pool properties indicating that references to objects in that pool need to be send a refcount.release op (or whatever)?  That would also apply to a simple redirect manifest if the target pool has that property, I would imagine?